### PR TITLE
feat: prefer cosmic-term in terminal plugin

### PR DESCRIPTION
**Pull request by wash2**
fixes https://github.com/pop-os/cosmic-launcher/issues/82

### CODE REVIEW

**Review by mmstick — APPROVED** [review]

File: plugins/src/terminal/mod.rs [hunk 1]
```diff
@@ -118,6 +122,15 @@ impl App {
 fn detect_terminal() -> (PathBuf, &'static str) {
     use std::fs::read_link;
 
+    let cosmic_term = Path::new("/usr/bin/cosmic-term");
```
Comment (Drakulix) [inline plugins/src/terminal/mod.rs:125]: It's a little ugly to hardcode /usr/bin. This means installing cosmic-term in a local prefix, or /usr/local/bin or on nix will not trigger this behavior. That probably isn't tooooo bad, but maybe we should just bite the bullet and parse $PATH?

Comment (wash2) [reply plugins/src/terminal/mod.rs:125]: Ya, I was thinking so too, but the exiting method of selecting a terminal for the plugin also used /usr/bin. I can change it though.

Comment (jackpot51) [reply plugins/src/terminal/mod.rs:125]: There is code in cosmic-files to detect terminal apps. That may be useful.

Comment (wash2) [reply plugins/src/terminal/mod.rs]: I've adjusted it to work more like cosmic-files, but I wonder if all terminal apps will be compatible, or if I should ignore the other terminal apps found searching the desktop entries when cosmic-term isn't found. alacritty seems to only support `-e` or `--command`, for example.

**Review by mmstick — DISMISSED** [review]

File: plugins/src/terminal/mod.rs [hunk 2]
```diff
@@ -116,8 +116,42 @@ impl App {
...
+    for id in ["com.system76.CosmicTerm"] {
+        for (terminal_id, exec) in &terminal_apps {
+            if terminal_id.as_str() == id {
+                return (PathBuf::from(exec), "--");
+            }
+        }
+    }
+
+    if let Some((_, exec)) = terminal_apps.first() {
+        return (PathBuf::from(exec), "--");
```
Comment (jackpot51) [inline plugins/src/terminal/mod.rs:152]: -- is not guaranteed to work. See xdg-terminal-exec, it recommends -e unless another argument is specified.

Comment (jackpot51) [reply plugins/src/terminal/mod.rs:152]: https://github.com/Vladimir-csp/xdg-terminal-exec?tab=readme-ov-file#desktop-entry-for-a-terminal

File: plugins/src/terminal/mod.rs [hunk 3]
```diff
@@ -116,8 +116,47 @@ impl App {
...
+    for id in ["com.system76.CosmicTerm"] {
+        for (terminal_id, exec) in &terminal_apps {
+            if terminal_id.as_str() == id {
+                return (PathBuf::from(exec), "-e");
+            }
+        }
+    }
+
+    if let Some((id, exec)) = terminal_apps.first() {
+        let arg = if id == "org.gnome.Terminal" {
```
Comment (jackpot51) [inline plugins/src/terminal/mod.rs:152]: gnome-terminal also supports `-e`, for now.

Comment (jackpot51) [reply plugins/src/terminal/mod.rs:152]: This is probably good enough for now, we might be able to push xdg-terminal-exec over the finish line.

Comment (wash2) [reply plugins/src/terminal/mod.rs:152]: Ya, I just saw that "-e" is deprecated for gnome-terminal, so I figured it might be best to do "--" instead

**Review by jackpot51 — APPROVED** [review]

**Comment by XV-02** [comment]
Nothing I have tried has failed yet. Are there any specific use cases that I should go out of my way to test?

**Comment by wash2** [comment]
> Nothing I have tried has failed yet. Are there any specific use cases that I should go out of my way to test?

If you haven't tried without cosmic-term installed, but some other terminal like alacritty or konsole that might be good.

**Pull request merged**